### ISSUE TITLE
Component modifications are not applied in module federation

### ISSUE THREAD

**Issue by kabel2** [pmmmwh/react-refresh-webpack-plugin#394]
I'm having an issue getting react refresh working on a project with module fedration. The ws connection is working, also in the console it seems its working, but the modifications are not applied to the component. 

```
[webpack-dev-server] App updated. Recompiling...
index.js:3210 [webpack-dev-server] App hot update...
log.js:24 [HMR] Checking for updates on the server...
log.js:24 [HMR] Updated modules:
log.js:24 [HMR]  - ./src/Button.js
log.js:24 [HMR] App is up to date.
```

I have created an basic example: https://github.com/kabel2/module-federation-examples/tree/feature/example-react-refresh/basic-host-remote

Changing the innertext of the button in the "Button.js" component in "app2" does not apply the changes live, only after reload. 

With the old react-hot-loader it seems to work.

**Comment by Ridermansb** [comment]
Same with me

**Comment by pmmmwh** [comment]
Are you externalising React?

**Comment by kabel2** [comment]
React is shared from App1 via the ModuleFederation plugin. I tried remove react from the list of shared components, but the problem remains.

**Comment by Ridermansb** [comment]
Any news on this?

**Comment by pmmmwh** [comment]
No. This is likely due to externalising and injection order of stuff. I currently do not have much time to check this, but I can revisit this later.

**Comment by Muritavo** [comment]
There are 2 key events for it to work correctly. react-refresh should be injecting before react-dom and also react-refresh should be executed only once, or if executed more than once (it seems like the case on app 2), the react-dom should be injected again.

With this piece of code at the start of your app and it should work right now. A PR would be welcome to fix it.

```typescript
const {injectIntoGlobalHook} = require("react-refresh/cjs/react-refresh-runtime.development")
injectIntoGlobalHook(window);
(window as any).__REACT_DEVTOOLS_GLOBAL_HOOK__.inject(ReactDOM);
```

**Comment by pmmmwh** [comment]
Maybe you could use the `library` option to ensure react refresh is injected for both apps?

**Comment by ScriptedAlchemy** [comment]
HMR cannot work with federated modules, you can use `@module-federation/fmr` as an alternative solution. 

We would have to rewrite HMR graphs in webpack in order to achieve this. Webpack needs to be aware of all module and parent modules in the graph which means we would have to build a way for federated webpack graphs to be interconnected as well. Changing a module in a remote doesn't mean that the container knows where its being used inside a host.

**Comment by ije** [comment]
i figured out how resovle this in #516, maybe not a greate idea, but it works

**Comment by raine** [comment]
Does anyone have an example how react-refresh etc. should be loaded on the page in order for hot reload to work?

In my module federation setup, hot reload works occasionally, which suggests there's some indeterminism going on in the order that scripts are loading on the page.

The suggestion of bundling react-refresh separately, placing it in window and then making it external in MF libs does not seem to work.

**Comment by ccambournac** [comment]
> Maybe you could use the `library` option to ensure react refresh is injected for both apps?

Indeed. Seems like using Webpack 5's `output.uniqueName` helped me solve this same issue.

**Comment by pmmmwh** [comment]
I would be happy to curate examples where React Refresh can be used properly with MF if anyone can provide a small sample repo.

**Comment by ccambournac** [comment]
> I would be happy to curate examples where React Refresh can be used properly with MF if anyone can provide a small sample repo.

I'll definitely try to find time to do so @pmmmwh.

**Comment by zhangHongEn** [comment]
> 使用ModuleFederation注意要满足这些条件才可以热更新, 我这里有一个例子可以供参考
https://www.npmjs.com/package/react-refresh-umd

**Comment by ruanyl** [comment]
For anyone who runs into this issue, the solution from @ije and @danieloprado in https://github.com/pmmmwh/react-refresh-webpack-plugin/pull/516 works perfectly for me. 

I end up with `pnpm patch` to patch the package, you can also do it with `yarn patch` if you're using `yarn`.

@pmmmwh would you reconsider applying the changes suggested by @ije and @danieloprado?

**Comment by ESoch** [comment]
As someone else who is experiencing issues here with module federation, I am hoping to re-up this conversation. 

@pmmmwh, would you be open to applying the changes described here: https://github.com/pmmmwh/react-refresh-webpack-plugin/pull/516#issuecomment-1159038683

As module federation becomes more popular I can imagine this will become a more common issue.

**Comment by ESoch** [comment]
Although, to further complicate things, I am not able to get it working even with the suggested fixes. I continue to get this error: 
```
   [HMR] Update failed: ChunkLoadError: Loading hot update chunk main failed.
```

@ruanyl, does your patch continue to work as expected for you?

**Comment by danieloprado** [comment]
@ESoch try disable runtimeChunk during development

```js
//...
optimization: {
    runtimeChunk: isDevelopment ? 'single' : false
 }
```

**Comment by zhangHongEn** [comment]
If you want to configure react-refresh to shared, you need react-refresh-webpack-plugin to change the import method of source code from react-refresh/runtime to react-refresh, and you also need to configure "react-refresh-webpack-plugin" include and exclude to avoid shared being referenced in the entry chunk, even so there will be other problems, so I also tend to force react-refresh to become a singleton to solve it quickly

**Comment by pmmmwh** [comment]
> As someone else who is experiencing issues here with module federation, I am hoping to re-up this conversation.
> 
> @pmmmwh, would you be open to applying the changes described here: [#516 (comment)](https://github.com/pmmmwh/react-refresh-webpack-plugin/pull/516#issuecomment-1159038683)
> 
> As module federation becomes more popular I can imagine this will become a more common issue.

I cannot really offer much help without any example to look into. For the patch, my stance still stands that polluting global and binding this package to browser only is a bad thing to do. I personally do not use module federation and is not knoledgeable with it by any means, so help would be appreciated.

**Comment by zhangHongEn** [comment]
> > As someone else who is experiencing issues here with module federation, I am hoping to re-up this conversation.
> > @pmmmwh, would you be open to applying the changes described here: [#516 (comment)](https://github.com/pmmmwh/react-refresh-webpack-plugin/pull/516#issuecomment-1159038683)
> > As module federation becomes more popular I can imagine this will become a more common issue.
> 
> I cannot really offer much help without any example to look into. For the patch, my stance still stands that polluting global and binding this package to browser only is a bad thing to do. I personally do not use module federation and is not knoledgeable with it by any means, so help would be appreciated.

@pmmmwh  You can refer to this online example, app2 achieves hmr by mounting react-refresh/runtime to the global maintain singleton, app3's react-refresh/runtime is not a singleton so it cannot hmr
[try online](https://stackblitz.com/github/wpmjs/examples/tree/main/module-federation-react-hmr)

**Comment by alexander-akait** [comment]
@felixmokross Maybe you can provide small reproducible example? It helps, thank you

**Comment by dtyyy** [comment]
> Maybe you could use the `library` option to ensure react refresh is injected for both apps?

``

**Comment by ScriptedAlchemy** [comment]
module federation v2 supports hmr now. https://module-federation.io/

**Comment by russellfoster** [comment]
> module federation v2 supports hmr now. https://module-federation.io/

Nice! Is there a way to get v2 in webpack that works with this plugin?

**Comment by ScriptedAlchemy** [comment]
Replace webpacks old plugin with the new one via PR, or use the official v2 that's not coupled to webpack.

**Comment by russellfoster** [comment]
> Replace webpacks old plugin with the new one via PR, or use the official v2 that's not coupled to webpack.

Thanks for responding. By the official v2 plugin, are you talking about "@module-federation/enhanced"?

https://www.npmjs.com/package/@module-federation/enhanced

**Comment by ScriptedAlchemy** [comment]
Yes that's the one

**Comment by jedwards1211** [comment]
If our host app is mounting React components from secondary MF container with the React instance bundled and shared by the host app MF container, should we also put `react-refresh/runtime` in the MF shared dependencies?
